Architecture.md

```diff
@@ -0,0 +1,287 @@
+prepared script everyone agrees on. Here is an example, assuming the trusted Admin is "John":
+
+* Init `x/validator` with Admin=John, Authority=nil, and an InitialValidator of one node
+* Upload group and multisig code (store code ids)
+* Init CW4/CW7-group contract with all pubkeys of "genesis" validators and proper group weights, Admin=John
+* Create both CW3-voting contracts pointing to the CW4/7-group contract (with different thresholds)
+* Set CW4/7-group admin to the voting contract #1 (eg. 60%)
+* Set `x/validator` Authority to the CW4/7-group contract (properly setting the validator set)
+* Set `x/validator` Admin to the voting contract #2 (eg. 80%)
+
+At this point, John no longer has any special access and all of the components are properly wired up together.
+
+My first step would to write such logic off-chain with CosmJS. We can then look how to include such logic
+on-chain. We could do this with [some init callbacks](https://github.com/CosmWasm/cosmwasm/issues/467),
+and uploading a smart contract with all this init logic as part of genesis. We could also do it with
+genesis-only Go code that parses genesis data (eg. pubkeys and voting weights) and runs this series of events.
+
+## Simple PoS Contract
+
+Once we have developed a PoA system using CosmWasm and demonstrated it working and properly updating the validators
+on a testnet, we can develop a simple PoS example (and swap them out live). The first PoS example only includes
+self-stake and doesn't have rewards or slashing, but let's show how this could work.
+
+Fundamentally, this will be quite similar to the PoA model above, with one change - the contract that updates
+the group. We will have:
+
+* CW4/7 group to hold current validator set in CosmWasm (as above)
+* Voting contract using this group that can serve as admin for `x/validator`
+* Possibly second voting contract for other governance (as needed)
+* NEW: token staking contract that is group admin and sets the member weights
+
+The staking contract must define a "staking token" (native or cw20) upon init, as well as an "unbonding period".
+It will have 3 main functionalities:
+
+* `SelfStake`: This locks the "staking token" sent with the message. It then adds the respective weight to the group contract.
+* `UnStake`: This reduces the sender's self-stake and creates an "undelegating token claim" with the proper expiration. It then updates the group by reducing the sender's weight.
+* `Release`: This checks all expired "undelegating token claims" by this validator and sends the appropriate tokens back to their wallet.
+
+## Distributing Rewards
+
+This requires an enhancement to the `x/validator` native module and a new CW8 reward distribution spec.
+We need to be able to mint block rewards
+in a native token, as well as distribute the FeeCollector tokens, which cannot be done with a generic CosmWasm contract
+(for obvious security reasons).
```

Comment:
What are the security reasons? 😳

Reply:
If we allow any contract mint block rewards or distribute the collected fees?